pkg/minikube/bootstrapper/kubeadm/kubeadm.go

```diff
@@ -0,0 +1,354 @@
+[Unit]
+Description=kubelet: The Kubernetes Node Agent
+Documentation=http://kubernetes.io/docs/
+
+[Service]
+ExecStart=/usr/bin/kubelet
+Restart=always
+StartLimitInterval=0
+RestartSec=10
+
+[Install]
+WantedBy=multi-user.target
+`
+
+const kubeadmConfigTmpl = `
+apiVersion: kubeadm.k8s.io/v1alpha1
+kind: MasterConfiguration
+api:
+  advertiseAddress: {{.AdvertiseAddress}}
+  bindPort: {{.APIServerPort}}
+kubernetesVersion: {{.KubernetesVersion}}
+certificatesDir: {{.CertDir}}
+networking:
+  serviceSubnet: {{.ServiceCIDR}}
+etcd:
+  dataDir: {{.EtcdDataDir}}
+nodeName: {{.NodeName}}
+`
+
+func NewKubeadmBootstrapper(api libmachine.API) (*KubeadmBootstrapper, error) {
+	h, err := api.Load(config.GetMachineName())
+	if err != nil {
+		return nil, errors.Wrap(err, "getting api client")
+	}
+	var cmd bootstrapper.CommandRunner
+	// The none driver executes commands directly on the host
+	if h.Driver.DriverName() == constants.DriverNone {
+		cmd = &bootstrapper.ExecRunner{}
+	} else {
+		client, err := sshutil.NewSSHClient(h.Driver)
+		if err != nil {
+			return nil, errors.Wrap(err, "getting ssh client")
+		}
+		cmd = bootstrapper.NewSSHRunner(client)
+	}
+	return &KubeadmBootstrapper{
+		c: cmd,
+	}, nil
+}
+
+//TODO(r2d4): This should most likely check the health of the apiserver
+func (k *KubeadmBootstrapper) GetClusterStatus() (string, error) {
+	statusCmd := `sudo systemctl is-active kubelet &>/dev/null && echo "Running" || echo "Stopped"`
+	status, err := k.c.CombinedOutput(statusCmd)
+	if err != nil {
+		return "", errors.Wrap(err, "getting status")
+	}
+	status = strings.TrimSpace(status)
+	if status == state.Running.String() || status == state.Stopped.String() {
+		return status, nil
+	}
+	return "", fmt.Errorf("Error: Unrecognized output from ClusterStatus: %s", status)
+}
+
+// TODO(r2d4): Should this aggregate all the logs from the control plane?
+// Maybe subcommands for each component? minikube logs apiserver?
+func (k *KubeadmBootstrapper) GetClusterLogs(follow bool) (string, error) {
+	var flags []string
+	if follow {
+		flags = append(flags, "-f")
+	}
+	logsCommand := fmt.Sprintf("sudo journalctl %s -u kubelet", strings.Join(flags, " "))
+
+	if follow {
+		if err := k.c.Run(logsCommand); err != nil {
+			return "", errors.Wrap(err, "getting shell")
+		}
+	}
+
+	logs, err := k.c.CombinedOutput(logsCommand)
+	if err != nil {
+		return "", errors.Wrap(err, "getting cluster logs")
+	}
+
+	return logs, nil
+}
+
+func (k *KubeadmBootstrapper) StartCluster(k8s bootstrapper.KubernetesConfig) error {
+	// We use --skip-preflight-checks since we have our own custom addons
+	// that we also stick in /etc/kubernetes/manifests
+	kubeadmTmpl := "sudo /usr/bin/kubeadm init --config {{.KubeadmConfigFile}} --skip-preflight-checks"
+	t := template.Must(template.New("kubeadmTmpl").Parse(kubeadmTmpl))
+	b := bytes.Buffer{}
+	if err := t.Execute(&b, struct{ KubeadmConfigFile string }{constants.KubeadmConfigFile}); err != nil {
```

Comment:
if you like, you can just use `map[string]string{"KubeadmConfigFile: constants.KubeadmConfigFile}` instead of the anonymous struct. it might read cleaner

Reply:
Ah that's a good idea. I'll do that